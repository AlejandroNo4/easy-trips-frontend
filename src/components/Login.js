import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import fetchingPost from '../api/fetchingPost';
import FormLogin from './FormLogin';
import fetchingGet from '../api/fetchingGet';

const Login = () => {
  const dispatch = useDispatch();
  const userState = useSelector((state) => state.UIReducer);
  const navigate = useNavigate();

  useEffect(() => {
    const url = 'logged_in';
    const type = 'UI';
    fetchingGet({
      dispatch,
      url,
      type,
    });
    if (userState.user.logged_in === true) navigate('/');
  }, []);

  const initialStateForm = {
    email: '',
    password: '',
  };
  const [form, updateInput] = useState(initialStateForm);

  const handleChange = (e) => {
    if (e.target.files) updateInput({ ...form, [e.target.name]: e.target.files[0] });
    else updateInput({ ...form, [e.target.name]: e.target.value });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    const {
      email, password,
    } = form;

    const formData = {
      user: {
        email: email.toLowerCase(),
        password,
      },
    };
    const url = '/sessions';
    const type = 'UI';
    fetchingPost({
      dispatch,
      url,
      formData,
      navigate,
      type,
    });
  };

  if (userState.loading === true) {
    return <h1>------LOADING...------</h1>;
  }
  return (
    <div>
      <FormLogin handleChange={handleChange} handleSubmit={handleSubmit} />
    </div>
  );
};

export default Login;
