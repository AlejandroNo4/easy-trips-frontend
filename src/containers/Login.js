import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useNavigate } from 'react-router-dom';
import fetchingPost from '../api/fetchingPost';
import FormLogin from '../components/FormLogin';
import fetchingGet from '../api/fetchingGet';
import { cleanupErrors } from '../actions';

const Login = () => {
  const dispatch = useDispatch();
  const userState = useSelector((state) => state.UIReducer);
  const navigate = useNavigate();

  useEffect(() => {
    const url = 'logged_in';
    const type = 'UI';
    fetchingGet({
      dispatch,
      url,
      type,
    });
    if (userState.user.logged_in === true) navigate('/');
    return () => {
      dispatch(cleanupErrors());
    };
  }, []);

  const initialStateForm = {
    email: '',
    password: '',
  };
  const [form, updateInput] = useState(initialStateForm);

  const handleChange = (e) => {
    if (e.target.files) updateInput({ ...form, [e.target.name]: e.target.files[0] });
    else updateInput({ ...form, [e.target.name]: e.target.value });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    const { email, password } = form;

    const formData = {
      user: {
        email: email.toLowerCase(),
        password,
      },
    };
    const url = '/sessions';
    const type = 'UI';
    fetchingPost({
      dispatch,
      url,
      formData,
      navigate,
      type,
    });
  };

  if (userState.loading === true) {
    return (
      <div className="d-flex flex-column justify-center align-center w-100">
        <h1 className="session-title">Loading...</h1>
      </div>
    );
  }

  let errors;

  if (userState.loading === false) {
    errors = userState.errors.message;
  }
  return (
    <div className="bg-no-session d-flex justify-center flex-column align-center no-session-container">
      <p className="error-msg">{errors}</p>
      <h1 className="session-title">Login</h1>
      <p className="session-description text-center">
        Hello there! please Login and start looking for the perfect trip.
      </p>
      <FormLogin handleChange={handleChange} handleSubmit={handleSubmit} />
      <Link to="/" className="link-back">
        Go back
      </Link>
    </div>
  );
};

export default Login;
