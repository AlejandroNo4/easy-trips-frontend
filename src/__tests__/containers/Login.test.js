import React from 'react';
import { Provider } from 'react-redux';
import {
  render, screen, cleanup, fireEvent, act,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import axiosMock from 'axios';
import { BrowserRouter } from 'react-router-dom';
import store from '../../store';
import Login from '../../containers/Login';

afterEach(cleanup);

const renderWithRedux = (component) => ({
  ...render(
    <Provider store={store}>
      <BrowserRouter>{component}</BrowserRouter>
    </Provider>,
  ),
});

describe('Login Component', () => {
  beforeEach(() => {
    renderWithRedux(<Login />);
  });

  it('renders with Redux', async () => {
    const loginTitle = await screen.findByTestId('login-container');
    expect(loginTitle).toBeVisible();
  });

  it('matches the snapshot', async () => {
    const loginTitle = await screen.findByTestId('login-container');
    expect(loginTitle).toMatchSnapshot();
  });

  it('Renders email input', async () => {
    const email = await screen.findByPlaceholderText('Email Adress');
    expect(email).toBeVisible();
    userEvent.type(email, 'anemail@mail.com');
    expect(email).toHaveValue('anemail@mail.com');
    expect(email).not.toHaveValue('');
  });

  it('Renders password input', async () => {
    const password = await screen.findByPlaceholderText('password');
    expect(password).toBeVisible();
    userEvent.type(password, '123456');
    expect(password).toHaveValue('123456');
    expect(password).not.toHaveValue('');
  });

  it('Cannot render and cannot render data wothout input typed', async () => {
    axiosMock.post.mockResolvedValueOnce({ data: {} });
    const password = await screen.findByPlaceholderText('password');
    const email = await screen.findByPlaceholderText('Email Adress');
    const button = await screen.findByTestId('login-btn');
    expect(password).toBeVisible();
    expect(email).toBeVisible();
    expect(button).toBeVisible();
    await act(async () => {
      fireEvent.click(button);
    });
    const message = await screen.findByTestId('errors-msg');
    expect(message).toBeVisible();
  });

  it('Makes a call to fetching action when filling the fields', async () => {
    axiosMock.post.mockResolvedValueOnce({ data: { logged_in: false } });
    const password = await screen.findByPlaceholderText('password');
    const email = await screen.findByPlaceholderText('Email Adress');
    const button = await screen.findByTestId('login-btn');
    expect(password).toBeVisible();
    expect(email).toBeVisible();
    expect(button).toBeVisible();
    userEvent.type(email, 'anemail@mail.com');
    userEvent.type(password, '123456');
    fireEvent.click(button);
    expect(axiosMock.post).toHaveBeenCalledTimes(1);
    expect(axiosMock.post).not.toHaveBeenCalledTimes(0);
    await screen.findByTestId('login-btn');
  });
});
