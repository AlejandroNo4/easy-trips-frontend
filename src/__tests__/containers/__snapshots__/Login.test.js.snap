// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component matches the snapshot 1`] = `
<div
  class="bg-no-session d-flex justify-center flex-column align-center no-session-container"
  data-testid="login-container"
>
  <p
    class="error-msg"
    data-testid="errors-msg"
  >
    Cannot destructure property 'data' of 'request' as it is undefined.
  </p>
  <h1
    class="session-title"
  >
    Login
  </h1>
  <p
    class="session-description text-center"
  >
    Hello there! please Login and start looking for the perfect trip.
  </p>
  <form
    class="d-flex flex-column align-center"
  >
    <input
      class="form-input"
      id="email"
      name="email"
      placeholder="Email Adress"
      required=""
      type="email"
    />
    <input
      class="form-input"
      id="password"
      name="password"
      placeholder="password"
      required=""
      type="password"
    />
    <button
      class="session-btn white"
      data-testid="login-btn"
      type="submit"
    >
      Log In
    </button>
  </form>
  <a
    class="link-back"
    href="/"
  >
    Go back
  </a>
</div>
`;
